osu.Game.Rulesets.Osu/Edit/OsuToolboxComposite.cs

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Input.Events;
+using osu.Game.Graphics.Containers;
+using osu.Game.Graphics.UserInterface;
+using osu.Game.Overlays.Settings.Sections;
+using osu.Game.Rulesets.Edit;
+using osu.Game.Screens.Edit;
+using osuTK;
+using osuTK.Input;
+
+namespace osu.Game.Rulesets.Osu.Edit
+{
+    /// <summary>
+    /// A toolbox composite for osu!-specific controls.
+    /// </summary>
+    // todo: once catch supports distance spacing, the control here should move out to a base "DistancingRulesetToolboxComposite" class or something better.
+    public class OsuToolboxComposite : CompositeDrawable
+    {
+        private ExpandingToolboxContainer expandingContainer;
+        private ExpandableSlider<double, SizeSlider<double>> distanceSpacingSlider;
+
+        private readonly Bindable<double> distanceSpacing = new BindableDouble(1.0)
+        {
+            MinValue = 0.1,
+            MaxValue = 6.0,
+            Precision = 0.01,
+        };
+
+        public IBindable<double> DistanceSpacing => distanceSpacing;
+
+        private bool distanceSpacingScrollActive;
+
+        [Resolved]
+        private EditorBeatmap editorBeatmap { get; set; }
+
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            RelativeSizeAxes = Axes.Both;
+
+            InternalChild = expandingContainer = new ExpandingToolboxContainer
+            {
+                Anchor = Anchor.TopRight,
+                Origin = Anchor.TopRight,
+                Child = new EditorToolboxGroup("snapping")
+                {
+                    Child = distanceSpacingSlider = new ExpandableSlider<double, SizeSlider<double>>
+                    {
+                        Current = { BindTarget = distanceSpacing },
+                        KeyboardStep = 0.1f,
+                    }
+                }
+            };
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            distanceSpacing.Value = editorBeatmap.BeatmapInfo.DistanceSpacing;
+            distanceSpacing.BindValueChanged(v =>
+            {
+                distanceSpacingSlider.ContractedLabelText = $"D. S. ({v.NewValue:0.##x})";
+                distanceSpacingSlider.ExpandedLabelText = $"Distance Spacing ({v.NewValue:0.##x})";
+                editorBeatmap.BeatmapInfo.DistanceSpacing = v.NewValue;
+            }, true);
+        }
+
+        protected override bool OnKeyDown(KeyDownEvent e)
+        {
+            if (e.AltPressed && e.Key == Key.D && !e.Repeat)
```

Comment:
I'm not sure where this shortcut comes from, but it's not correct. You can't have Alt+Alpha+Mouse input, it breaks unspoken UX rules.

Reply:
For sure, this is open for suggestions. Stable uses Alt+Scroll for modifying the distance spacing, but as you know, that's a global-wide shortcut for modifying the game value.

Reply:
Yeah, I'm not yet sure on the best way to handle this. We already overload the mouse wheel too much. For the sake of keeping things moving forward, you could make it ctrl+alt for now?